### PR TITLE
[docs] Update README

### PR DESCRIPTION
**Pull request by abiogenesis-now**
@hausdorff @jessicayuen Would appreciate eyes and feedback on this even thought it's not 100% done. 

**TODOs:**
- [X] ~Figure out why `kuard` image from the quickstart script isn't actually showing anything (something to do with the port forwarding?)~ (swapped out with guestbook image)
- [X] Update README in `docs/cli-reference` to be more helpful (this is temporary until we figure out how to use this autogenerated markdown on the site)

Let me know if you have any questions. Hopefully the short annotations in the *Additional Documentation* help give a sense of where this is going.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,64 +1,58 @@
-## Managing vendor/
```
Comment (jessicayuen) [inline CONTRIBUTING.md]: Hmm, I don't think we should be removing the contents of the original file, which I found particularly helpful.

Maybe add the DCO Sign off as another section to this file?

File: CODE-OF-CONDUCT.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+comments, commits, code, wiki edits, issues, and other contributions that are not
+aligned to this Code of Conduct. By adopting this Code of Conduct, project maintainers
+commit themselves to fairly and consistently applying these principles to every aspect
+of managing this project. Project maintainers who do not follow or enforce the Code of
+Conduct may be permanently removed from the project team.
+
+This code of conduct applies both within project spaces and in public spaces
+when an individual is representing the project or its community.
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting a project maintainer, Alex Clemmer (alex@heptio.com) and TBD.
```
Comment (jessicayuen) [inline CODE-OF-CONDUCT.md]: I think it's sufficient to follow the Ark/Sonobouy model and mention `by contacting the project maintainer(s)`, so we don't have to constantly update this file.

Or - replace with the ksonnet / heptio ksonnet mailing list.

File: README.md [hunk 3]
```diff
@@ -6,75 +6,149 @@
...
-# Can also use explicit `-J` args everywhere.
-% export KUBECFG_JPATH=/path/to/ksonnet/lib
+Here we provide a shell script that shows some basic ksonnet features in action. You can run this script to deploy the [Kuard container image](https://github.com/kubernetes-up-and-running/kuard), via a Kubernetes Service and Deployment.
+
+**Minimal explanation is provided, and only basic features are shown---this is intended to be a quick demonstration.** If you are interested in learning more, see [Additional Documentation](#additional-documentation).
+
+### Prerequisites
+* *You should have access to an up-and-running Kubernetes cluster (**maximum** version 1.7).* If you do not have a cluster, [choose a setup solution](https://kubernetes.io/docs/setup/) from the official Kubernetes docs.
+
+* *You should have `kubectl` installed.* If not, follow the instructions for [installing via Homebrew (MacOS)](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-with-homebrew-on-macos) or [building the binary (Linux)](https://kubernetes.io/docs/tasks/tools/install-kubectl/#tabset-1)
```
Comment (jessicayuen) [inline README.md]: Missing `.` at end of line.

File: README.md [hunk 4]
```diff
@@ -6,75 +6,149 @@
...
+
+* *Your `$KUBECONFIG` should specify a valid `kubeconfig` file*, which points at the cluster you want to use for this demonstration.
+
+### Script
+
+In one shell session:
+```
+# Start by creating your app directory'
```
Comment (jessicayuen) [inline README.md]: Accidental `'`?

File: README.md [hunk 5]
```diff
@@ -6,75 +6,149 @@
...
+
+ksonnet is a feature-rich framework. To learn more about how to integrate it into your workflow, check out the resources below:
+
+* **Tutorial (WIP)** - How do I use ksonnet and why?
+
+* **Interactive tour of ksonnet (WIP)** - Where does the ksonnet magic come from?
+
+* **[CLI Reference](/docs/cli-reference)** - What ksonnet commands are available, and how do I use them?
+
+* **Concept Reference(WIP)** - What do all these special ksonnet terms mean (e.g. *prototypes*) ?
```
Comment (jessicayuen) [inline README.md]: space before `(`

**Comment by jessicayuen** [comment]
Do you think you can squash some of these commits before merging? Ex: address comments

**Comment by abiogenesis-now** [comment]
yeah, I've been changing a fair amount, will rebase once they're reviewed---
@jessicayuen, do you mind taking a look at the two most recent commits? (`address comments` and `update...`)

TLDR of what's changed:
* `CONTRIBUTING.md`
* changed quickstart script to use guestbook image instead of kuard image (https://heptio.slack.com/archives/C49LH42KD/p1510844821000483)
* added some stuff to the CLI readme (linked from main README)

If you don't mind, I'll make a followup PR to add the `concepts.md` stuff, just so we can get the updated README merged first. let me know if you have any questions, thanks!

File: README.md [hunk 6]
```diff
@@ -6,75 +6,168 @@
...
 ## Quickstart
 
-```console
-# Include <ksonnet.git>/lib in ksonnet/jsonnet library search path.
-# Can also use explicit `-J` args everywhere.
-% export KUBECFG_JPATH=/path/to/ksonnet/lib
+Here we provide a shell script that shows some basic ksonnet features in action. You can run this script to deploy and update a basic web app UI, via a Kubernetes Service and Deployment. This app is shown below:
+
+<p align="center">
+<img alt="guestbook screenshot" src="/docs/img/guestbook.png" style="width:60% !important;">
```
Comment (jessicayuen) [inline README.md]: Nit: Can we remove the "af", "blah", etc. words from this screenshot? Or update it to "Hello, ksonnet" or something.

File: README.md [hunk 7]
```diff
@@ -6,75 +6,168 @@
...
+* *Your `$KUBECONFIG` should specify a valid `kubeconfig` file*, which points at the cluster you want to use for this demonstration.
 
-# Update to new config
-% ks apply -f examples/guestbook.jsonnet
+### Script
+
+Copy and paste the script below to deploy the container image for a basic web app UI:
+
+```
+# Start by creating your app directory
```
Comment (jessicayuen) [inline README.md]: I do think it's worth mentioning here that we're initializing the app based on the current context in their configurations.

File: README.md [hunk 8]
```diff
@@ -6,75 +6,168 @@
...
+# Autogenerate a basic manifest
+ksonnet generate deployment-exposed-with-service guestbook-ui \
+--name guestbook \
+--image alpinejay/dns-single-redis-guestbook:0.3 \
+--type ClusterIP
+
+# Deploy your manifest to your cluster
+ksonnet apply default
+
+# Set up an API proxy so that you can access the guestbook service locally
```
Comment (jessicayuen) [inline README.md]: What's your opinion on removing this entirely and asking them to visit their server? I feel that they should have this information based on the pre-reqs.

File: README.md [hunk 9]
```diff
@@ -6,75 +6,168 @@
...
+```
+
+Even though you've made modifications to the Guestbook app and removed it from your cluster, ksonnet still tracks all your manifests locally:
+
+```
+# View all expanded manifests (YAML)
+ks show default
+```
+
+If you're wondering how ksonnet differs from existing tools, the [full-length tutorial (WIP)]() shows you how to use other ksonnet features to implement the rest of the Guestbook app (so that the buttons work!).
```
Comment (jessicayuen) [inline README.md]: The full-length tutorial goes to a missing page. Is this intentional or is there supposed to be a stub (i.e. to a TODO page similar to concepts)?

File: docs/concepts.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+### Application
+TODO
+
+### Component
+TODO
+
+### Prototype
+TODO
+
+### Parameter
```
Comment (jessicayuen) [inline docs/concepts.md]: I know you're going to be updating this in a separate PR, but just wanted to mention that `Environment` should probably be one of these concepts.

File: README.md (hunk 6 above)
Comment (abiogenesis-now) [reply README.md]: good catch

File: README.md (hunk 8 above)
Comment (abiogenesis-now) [reply README.md]: I want to avoid blocking users who aren't familiar with services/bastion hosts/etc. `LoadBalancer` doesn't work on all clusters (only AWS/GCP/Azure, *and* it's billable), so I opted for API proxy as a relatively agnostic way of getting them to the webpage

File: README.md (hunk 9 above)
Comment (abiogenesis-now) [reply README.md]: yes, the tutorial will be hosted on the new ksonnet site, and is not yet completed. if it is easier (and less confusing), I can remove the link for now

File: README.md (hunk 7 above)
Comment (abiogenesis-now) [reply README.md]: good point

File: docs/concepts.md (hunk 10 above)
Comment (abiogenesis-now) [reply docs/concepts.md]: ahh right, thanks!

**Comment by abiogenesis-now** [comment]
@jessicayuen I addressed most of your comments (and left a response on the one I didn't change). PTAL, thank you!

File: CONTRIBUTING.md (hunk 1 above)
Comment (hausdorff) [reply CONTRIBUTING.md]: Just so we're all clear: we actually shouldn't remove the contributing file, as we need it for scripts.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,64 +1,96 @@
...
 
-This is a good one to run before sending a non-trivial change for
-review.
+### Build
 
+To build ksonnet locally (e.g. for open-source development), run the following in the root of your ksonnet repo:
 ```
-# See unused
-govendor list +unused
+make all
```
Comment (hausdorff) [inline CONTRIBUTING.md]: Are we calling out a dependency on `GOPATH` here? I must be missing it.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,64 +1,96 @@
...
 
-# Remove unused
-govendor remove +unused
+*Troubleshooting*
+
+`make docs` relies on the `realpath` Linux utility. If you encounter a `realpath: command not found` error, and you are using OSX, run the following command:
+```
+brew install coreutils
 ```
+Running `make` again afterwards should work.
```
Comment (hausdorff) [inline CONTRIBUTING.md]: Might want to call out `make test` here too if this is the contributing doc.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,64 +1,96 @@
...
+
+### Manage dependencies
 
-### Add a new dependency
+This project uses `govendor` to manage Go dependencies. To make things easier for reviewers, put updates to `vendor/` in a separate commit within the same PR.
```
Comment (hausdorff) [inline CONTRIBUTING.md]: I'd specifically call out the requirements of adding a vendored dependency. Perhaps something like (this is a really quick cut, feel free to discard):

```
This project uses `govendor` to manage dependencies. When you introduce a dependency to the ksonnet codebase, you need to:

1. Discuss the use case and need for the dependency with the maintainers by opening an issue.
1. After building agreement, vendor the dependency using `govendor`.
  1. Where possible, **please pin the repository to a specific stable version.** Please do not vendor HEAD if you can avoid it!
  1. **Please introduce a vendored dependency in a dedicated commit to make review easier!**
1. Write the code that exercises the dependency.
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,64 +1,96 @@
...
 
 ```
-# See missing
+# View missing dependencies
 govendor list +missing
 
-% govendor fetch -v +missing
+# Download missing dependencies into vendor/
+govendor fetch -v +missing
 ```
```
Comment (hausdorff) [inline CONTRIBUTING.md]: Following from the above comment, I'd also mention that you should check in the changes to `vendor` separately from your new app code.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,64 +1,96 @@
...
+1. Add an appropriate tag.  We do this via `git` (not the github UI) so that the tag is signed.  This process requires you to have write access to the real `master` branch (not your local fork).
    ```
-   % tag=vX.Y.Z
-   % git fetch   # update
-   % git tag -s -m $tag $tag origin/master
-   % git push origin tag $tag
+   tag=vX.Y.Z
+   git fetch   # update
+   git tag -s -m $tag $tag origin/master
+   git push origin tag $tag
```
Comment (hausdorff) [inline CONTRIBUTING.md]: I believe this is handled if you do something like `VERSION=0.7 make`?

File: README.md [hunk 16]
```diff
@@ -6,75 +6,169 @@
...
+* **Customize** manifests directly with powerful object concatenation syntax
+* **Deploy** app manifests to multiple environments
+* **Diff** across environments to compare two running versions of your app
+* **Track** the entire state of your app configuration in version controllable files
+
+All of this results in a more iterative process for developing manifests, one that can be supplemented by continuous integration (CI).
 
 *STATUS: Development is ongoing—this tool is pre-alpha.*
 
 ## Install
```
Comment (hausdorff) [inline README.md]: Probably should make a note to add a "releases" section when we cut a release, too.

**Comment by abiogenesis-now** [comment]
right it should still be there

File: README.md [hunk 17]
```diff
@@ -6,75 +6,169 @@
...
 
-To build from source:
+> You should have Go installed, and an appropriately set `$GOPATH`. (For most systems, the minimum Go version is 1.7. However, recent OSX may [require golang >=1.8.1](https://github.com/golang/go/issues/19734) to
```
Comment (hausdorff) [inline README.md]: We support only > 1.8 actually. We're using stuff like `sort.Slice`.

File: README.md [hunk 18]
```diff
@@ -6,75 +6,169 @@
...
+avoid an immediate `Killed: 9`.) If you need additional help, see the [official Go installation guide](https://golang.org/doc/install#install).
+
+ksonnet is installed like any other Go binary. Run the following shell commands to download ksonnet and ensure that it is runnable in any directory:
```
Comment (hausdorff) [inline README.md]: How about something like: `ks` is installed like any other Go binary.

File: README.md [hunk 19]
```diff
@@ -6,75 +6,169 @@
...
+
+```bash
+go get github.com/ksonnet/ksonnet
+PATH=$PATH:$GOPATH/bin
+```
+
+If your ksonnet is properly installed, you should be able to run the following `--help` command and see similar output:
```
Comment (hausdorff) [inline README.md]: Might want to standardize on calling the binary `ks` to avoid confusion.

File: README.md (hunk 6 above)
Comment (hausdorff) [reply README.md]: lol I kind of like "af", but if we want to be all "professional" I understand.

File: README.md [hunk 20]
```diff
@@ -6,75 +6,169 @@
...
+
+```
+# Start by creating your app directory
+# (This references your current cluster using $KUBECONFIG)
+ksonnet init quickstart
+cd quickstart
+
+# Autogenerate a basic manifest
+ksonnet generate deployment-exposed-with-service guestbook-ui \
+--name guestbook \
```
Comment (hausdorff) [inline README.md]: Might want to consider indenting these flags. Like:

```
ksonnet generate deployment-exposed-with-service guestbook-ui \
  --name guestbook \
  --image alpinejay/dns-single-redis-guestbook:0.3 \
  --type ClusterIP
```

File: README.md (hunk 16 above)
Comment (abiogenesis-now) [reply README.md]: ^ this is in the *Changelog* section at the bottom of the README

**Comment by abiogenesis-now** [comment]
@hausdorff addressed all your comments, PTAL when you get a chance, thanks!

**Comment by abiogenesis-now** [comment]
^ left it in a separate commit for now so changes are easier for you to review
